src/rgw/rgw_rest_bucket_logging.cc

```diff
@@ -111,54 +117,67 @@ class RGWPutBucketLoggingOp : public RGWDefaultResponseOp {
-      op_ret = -EINVAL;
-      return;
+    if (const auto ret = verify_bucket_logging_params(this, s); ret < 0) {
+      return ret;
     }
 
-    RGWXMLDecoder::XMLParser parser;
-    if (!parser.init()){
-      ldpp_dout(this, 1) << "ERROR: failed to initialize XML parser" << dendl;
-      op_ret = -EINVAL;
-      return;
+    {
+      const auto max_size = s->cct->_conf->rgw_max_put_param_size;
+      bufferlist data;
+      int ret = 0;
+      if (std::tie(ret, data) = read_all_input(s, max_size, false); ret < 0) {
+        ldpp_dout(this, 1) << "ERROR: failed to read XML logging payload, ret = " << ret << dendl;
+        return ret;
+      }
+      if (data.length() == 0) {
+        ldpp_dout(this, 1) << "ERROR: XML logging payload missing" << dendl;
+        return -EINVAL;
+      }
+
+      RGWXMLDecoder::XMLParser parser;
+      if (!parser.init()){
+        ldpp_dout(this, 1) << "ERROR: failed to initialize XML parser" << dendl;
+        return -EINVAL;
+      }
+      if (!parser.parse(data.c_str(), data.length(), 1)) {
+        ldpp_dout(this, 1) << "ERROR: failed to parse XML logging payload" << dendl;
+        return -ERR_MALFORMED_XML;
+      }
+      configuration.default_obj_roll_time = get_cct()->_conf->rgw_bucket_logging_obj_roll_time;
+      try {
+        RGWXMLDecoder::decode_xml("BucketLoggingStatus", configuration, &parser, true);
+      } catch (RGWXMLDecoder::err& err) {
+        ldpp_dout(this, 1) << "ERROR: failed to parse XML logging payload. error: " << err << dendl;
+        return -ERR_MALFORMED_XML;
+      }
     }
-    if (!parser.parse(data.c_str(), data.length(), 1)) {
-      ldpp_dout(this, 1) << "ERROR: failed to parse XML logging payload" << dendl;
-      op_ret = -ERR_MALFORMED_XML;
-      return;
+
+    if (const auto ret = rgw::bucketlogging::get_bucket_id(configuration.target_bucket, s->bucket_tenant, target_bucket_id); ret < 0) {
+      ldpp_dout(this, 1) << "ERROR: failed to parse target bucket '" << configuration.target_bucket << "', ret = " << ret << dendl;
+      return ret;
     }
-    rgw::bucketlogging::configuration configuration;
-    configuration.default_obj_roll_time = get_cct()->_conf->rgw_bucket_logging_obj_roll_time;
-    try {
-      RGWXMLDecoder::decode_xml("BucketLoggingStatus", configuration, &parser, true);
-    } catch (RGWXMLDecoder::err& err) {
-      ldpp_dout(this, 1) << "ERROR: failed to parse XML logging payload. error: " << err << dendl;
-      op_ret = -ERR_MALFORMED_XML;
-      return;
+
+    if (const auto ret = driver->load_bucket(this, target_bucket_id,
+                                 &target_bucket, y); ret < 0) {
+      ldpp_dout(this, 1) << "ERROR: failed to get target bucket '" << target_bucket_id << "', ret = " << ret << dendl;
+      return ret;
     }
 
-    const rgw_bucket src_bucket_id(s->bucket_tenant, s->bucket_name);
+    return rgw::bucketlogging::verify_target_bucket_policy(this, target_bucket.get(), target_bucket_id, configuration, s);
```

Comment:
we discussed in slack why it was ok for `log_record()` to do the target bucket permission checks outside of `verify_permission()` without admin override. maybe the same applies to this check on PutBucketLogging?

but as-is, relying on `verify_permission()` to load `target_bucket` and parse the `BucketLoggingStatus` is problematic: if called by an admin that isn't the bucket owner, the `is_owner_of` check at the top will return an error, but the admin override will go on to call `execute()` with uninitialized `target_bucket` and `configuration`

if necessary, you can override `init_processing()` for that loading/parsing step which happens before the call to `verify_permission()` - and any errors from `init_processing()` will fail the request without admin override